garden_ai/garden.py

```diff
@@ -79,3 +81,65 @@ def _authenticate(self):
             on_refresh=self.auth_key_store.on_refresh,
         )
         return authorizer
+
+    def create_garden(self, **kwargs):
+        """Initialize and return a new Garden object
+
+        If invoked with keyword arguments corresponding to field names (e.g.
+        `title` or `creator`) are passed directly to the `Garden` model, making
+        it possible for the user to provide necessary info upfront to be parsed
+        and validated at initialization, or to incrementally provide field
+        values via simple assignment to attributes of the returned object
+        directly.
+
+        Parameters
+        ----------
+        **kwargs :
+            Metadata for the new Garden object. Keyword arguments matching
+            required or recommended fields will be coerced to the appropriate
+            type, if necessary.
+
+        Examples
+        --------
+
+            gc = GardenClient()
+            pea_garden = gc.create_garden(
+                creators=["Mendel, Gregor"], title="Experiments on Plant Hybridization"
+            )
+            pea_garden.year = 1863
+        """
+        return Garden(**kwargs)
+
+    def register_garden(self, G: Garden):
+        """Finalize the setup of a Garden object with user-populated fields
+
+        Parameters
+        ----------
+        G : Garden
+
+        Raises
+        ------
+        TypeError
+
+        Examples
+        --------
+        TODO
+        """
+        def request_doi(garden):
+            # TODO
+            # to be replaced with code to build the actual doi
+            return "10.55555/fake-doi"
+        try:
+            # attempts to populate a dummy garden with the data from the real
+            # one, so that it invokes our custom validators on every field (in
+            # particular, those which still have TempField defaults)
+            G.doi = request_doi(G)
+            _ = Garden(**G.dict())
+        except ValidationError as e:
+            logger.error(str(e))
+            print(e)
+            raise TypeError
+        else:
+            outfile = Path.cwd() / 'metadata.json'
+            with open(outfile, 'w+') as f:
+                f.write(G.json())
```

Comment:
what's the purpose of this? (this is a geniuine question, not a review comment)

Reply:
just to hit the acceptance criteria spelled out in #4